Adding the ContainerAwareAuthenticationProviderManager to lazily load…

weaverryan opened the pull request:
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | yes [minor, see below] |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | doctrine/DoctrineBundle#351 |
| License | MIT |
| Doc PR | n/a |

Hi guys!

**Problem** During the cache warming process, a huge chain takes place in the container: Twig->AuthorizationChecker->AuthenticationManager->Authentication Providers -> User Provider -> Doctrine -> ... potentially trigger a database look up :/

This doesn't entirely solve the linked issue, but it does solve a very common case that's caused when you use FOSUserBundle. The same could be done to make the voters lazy in `AccessDecisionManager`.

Do we like this approach? I was taking a shot a breaking the dependency chain.

**BC Breaks**
There are a few BC breaks, potentially

1) (very minor) If you don't have any providers configured or one does not implement the correct interface, the exception is thrown a a different _time_ (not when the object is instantiate, but when authenticate is called). This could be fixed (we'd just have a little bit of duplication)

2) If you are using the FW and type-hinted against `AuthenticationProviderManager`, that will break. If you type-hint the interface, you're fine

Cheers!

weaverryan committed:
Adding the ContainerAwareAuthenticationProviderManager to lazily load auth providers

weaverryan committed:
fabbot - you stickler!

iltar commented:
Shouldn't the correct "circuit breaker" be the twig part loading the authorization checker? This change will have a big impact where as I can imagine that loading a twig extension with the container instead of the service, is a small impact.

weaverryan commented:
I was also considering that too. But if the lazy loading of the providers is something we think is "good" on general, then this would be a better fix (it also fixed a *bit* more cases - i.e. for other cache warmers that might need the security system). 

But, breaking it at the Twig Extension is easier, just not as effective. So, I'm mixed

iltar commented on src/Symfony/Component/Security/Core/Authentication/ContainerAwareAuthenticationProviderManager.php:
```diff
@@ -0,0 +1,50 @@
+        $providers = array();
+        foreach ($this->providerServiceIds as $serviceId) {
+            $providers[] = $this->container->get($serviceId);
+        }
```
Maybe this list should be cached internally if getProviders can be called more often. The other option would be to yield the providers. This solution would mean that the earlier it finds a provider, the less get calls to the container are executed (if I understand yield properly that is).

weaverryan replied:
yield, brilliant! :)

> (if I understand yield properly that is)

Don't you have at least one yield in core? ;)

iltar replied:
Yes I do, but I mean more of the internal working. My understanding of yield:
 1. Goes through the generator method until it finds a yield
 2. Then goes into the foreach until that iteration is done
 3. Goes back into the generator method until it finds a yield
 4. goto 2

This is how I assume it works because it also reduces memory usage by not having to build an array, hence I believe it would be a benefit to use here (for both memory and method calls) as it's always used in an iteration.

Downside, `if (!$providers) {` can't be used. The only way to determine if there's no providers, is to see if the foreach was triggered or not.

sstok commented on src/Symfony/Component/Security/Core/Authentication/AbstractAuthenticationProviderManager.php:
```diff
@@ -0,0 +1,116 @@
+            throw new \InvalidArgumentException('You must at least add one authentication provider.');
+        }
+
+        foreach ($providers as $provider) {
```
Would you rather add an option (debug/verbose) to use eg. two loops (one for checking and one for actual operation) or one loop (check during operation)?

weaverryan replied:
good idea (to check in one loop) - I hadn't seen that I have 2 loops here!

weaverryan committed:
Checking the instances in the loop, to avoid 2 loops

This has *slightly* different behavior, but since we're just doing
checking for mistakes made by the developer, I think that's ok.

weaverryan committed:
Using yield for a slight performance increase

weaverryan commented:
Closing - I realized this PR is impossible, as currently it is ok for auth providers to be registered as private services, and this requires them all to be public. That's not a BC break we can fix in any way with this approach :/
